### PR TITLE
Mam test

### PR DESCRIPTION
**Pull request by thomleysens**
# Details on UDV-Core modifications

### Summary

| Issue | State | iTowns issues |
|:------|:------|:--------------|
| Add and configure vectors layers (*lines, polygons, points*) only from JSON file in order to avoid modifications in JavaScript and simplify the process | Working, bug on ```linewidth``` and multipolygons | https://github.com/iTowns/itowns/issues/1182 https://github.com/iTowns/itowns/issues/1179 |

### Modified files and linked files
| Files | Type | Directory (*relative to UDV-Core directory*) | Status |
|:------|:-----|:--------------------------------------------:|-------:|
| BaseDemo.js | JS | */src/Utils/BaseDemo/js/* | ***Required*** |
| layers_settings.json | JSON | */examples/MAM/data/config/* | ***Required*** |
| mini_poly.geojson | GeoJSON | */examples/MAM/data/geojson/* | ***Optional*** |

#### Presentation and examples

The JSON configuration file is a list of JSON object structured like this:
* for ***WFS*** source (*example below*):
  ```JSON
  {
    "id": "Lignes Bus TCL",
    "geometry": "lines",
    "linewidth": 50,
    "source": {
      "url": "https://download.data.grandlyon.com/wfs/rdata?",
      "protocol": "wfs",
      "version": "2.0.0",
      "typeName": "tcl_sytral.tcllignebus",
      "projection": "EPSG:3946",
      "zoom": {
        "min": 2,
        "max": 2
      },
      "format": "geojson"
    },
    "color": {
      "red": 0.6,
      "green": 0.6,
      "blue": 1.0
    }
  }
  ```
* for ***File*** source (*example below*):
  ```JSON
  {
    "id": "Lyon mask 10mns",
    "geometry": "polygons",
    "source": {
      "format": "geojson",
      "url": "http://localhost:8080/examples/MAM/data/geojson/mini_poly.geojson",
      "projection": "EPSG:3946",
      "zoom": {
        "min": 2,
        "max": 2
      }
    },
    "color": {
      "red": 1.0,
      "green": 0.6,
      "blue": 1.0
    }
  }
  ```

Let see some of the main parameters in details.

#### Parameters in details

We use parameters allowed by iTowns tools. We added some parameters (*geometry, color, ...*) to facilitate the management.

* ```id```(**str**): *unique id/title for the layer*
* ```geometry```(**str**): *possible values for the geometry*:
  * lines,
  * polygons,
  * points
* ```size```(**str**): *ONLY FOR POINTS [OPTIONAL], size of points*
* ```linewidth```(**str**): *ONLY FOR LINES [OPTIONAL], width of lines*
* ```source```(**json object**): *source of the data*:
  * ```format```(**str**): *format of the source, ex: "geojson"*,
  * ```url```(**str**): *url of the "local" source*,
  * ```protocol```(**str**): *protocol to use, **must be deleted for file source**. Ex: "wfs"*,
  * ```projection```(**str**): *EPSG of the source, ex: "EPSG:4326"*,
  * ```zoom```(**json object**): *range of zoom*:
    * ```min```(**int**): *zoom minimum*,
    * ```max```(**int**): *zoom maximum*
* ```color```(**json object**): *color of the elements*:
  * ```red```(**float**): *proportion of red, from 0.0 to 1.0*
  * ```green```(**float**): *proportion of green, from 0.0 to 1.0*
  * ```blue```(**float**): *proportion of blue, from 0.0 to 1.0*

This list of parameters will be read by JavaScript functions.

#### Set specific variables

First, we had to set specific variables (*altitude and color*) in order to simplify the process:

```JavaScript
let color = new itowns.THREE.Color();
    let rgb;

    // ********* Function to set altitude relative to tile altitude
    var altitude = (properties, contour) => {
      var result;
      var z = 0;
      if (contour) {
        result = itowns.DEMUtils.getElevationValueAt(this.view.tileLayer, contour, 0, tile);

        if (!result) {
          result = itowns.DEMUtils.getElevationValueAt(this.view.tileLayer, contour, 0);
        }
        tile = [result.tile];
        if (result) {
          z = result.z;
        }
        return z + 10;
      }
    };

    // ********* Set the absolute value for polygons layer altitude
    var altitudePoly = 200;
```

The altitude of polygons layers is set with the altitudePoly variable (*an integer*) for the moment. But this must be improved in order to manage altitude regarding the tile as in the lines/points case (*with the altitude function*).

Then we can parse the configuration JSON file and loop through all the JSON objects in the list.

#### Parse the JSON configuration file and loop through JSON objects

* Get the ```geometry```, the ```protocol``` (*if any*), and set global variables (```config```, ```size```, ```linewidth```)

```JavaScript
itowns.Fetcher.json('../../../examples/MAM/data/config/layers_settings.json')
      .then(
        (configList) => {
          for (var i in configList) {
            let geometry = configList[i].geometry;
            let protocol = configList[i].source.protocol;
            let config;
            let size;
            let linewidth;
```

* Set the parameters (*default if undefined in the parameters file*) for each case within a unique function:
  * points,
  * polygons,
  * lines (***WARNING: linewidth doesn't work for now***)

```JavaScript
let addGeometryLayerFromConfig = (config) => {
  if (geometry === "polygons"){
    config.altitude = altitudePoly
  }
  else if (geometry === "lines"){
    config.altitude = altitude;
    if (config.linewidth){
      linewidth = config.linewidth
    }
    else {
      linewidth = 50
    };
    config.onMeshCreated = function(result) {
      result.traverse(function _setLineWidth(mesh) {
        if (mesh.material) {
          mesh.material.linewidth = linewidth;
        }
      });
    }
  }
  else if (geometry === "points"){
    config.altitude = altitude;
    if (config.size){
      size = config.size
    }
    else {
      size = 50
    };
    config.onMeshCreated = function(result) {
      result.traverse(function _setPointSize(mesh) {
        if (mesh.material) {
          mesh.material.size = size;
        }
      });
    }
  };
  config.update = itowns.FeatureProcessing.update;
  config.convert = itowns.Feature2Mesh.convert({
    color: new itowns.THREE.Color().setRGB(
      config.color.red,
      config.color.green,
      config.color.blue
    ),
    altitude: config.altitude,
  });
  config.overrideAltitudeInToZero = true;
  config.projection = config.source.projection;
  config.source = config.source;
  let layer = new itowns.GeometryLayer(
    config.id,
    new itowns.THREE.Group(),
    config
  );
  return this.view.addLayer(layer);
};
```

* Set the ```configuration.source``` regarding the ```protocol```:

```JavaScript
if (protocol === undefined) {
  let configuration = configList[i];
  var promise = new Promise(
    function(resolve, reject) {
      configuration.source = new itowns.FileSource(configList[i].source);
      resolve(configuration);
    }
  );
  promise.then(addGeometryLayerFromConfig);
}

//// If protocol === "wfs", set configuration
else if (protocol === "wfs") {
  let configuration = configList[i];
  var promise = new Promise(
    function(resolve, reject) {
      configuration.source = new itowns.WFSSource(configList[i].source);
      resolve(configuration);
    }
  );
  promise.then(addGeometryLayerFromConfig);
}
```

### Complete source code for these functions

```JavaScript
// ********* COLOR AND ALTITUDE BASIC SETTINGS
    let color = new itowns.THREE.Color();
    let rgb;

    // ********* Function to set altitude relative to tile altitude
    var altitude = (properties, contour) => {
      var result;
      var z = 0;
      if (contour) {
        result = itowns.DEMUtils.getElevationValueAt(this.view.tileLayer, contour, 0, tile);

        if (!result) {
          result = itowns.DEMUtils.getElevationValueAt(this.view.tileLayer, contour, 0);
        }
        tile = [result.tile];
        if (result) {
          z = result.z;
        }
        return z + 10;
      }
    };

    // ********* Set the absolute value for polygons layer altitude
    var altitudePoly = 200;

    // ********* Function to get color for lines
    var colorLine = function(properties) {
      if (properties) {
        var rgb = properties.couleur.split(' ');
        return color.setRGB(rgb[0] / 255, rgb[1] / 255, rgb[2] / 255);
      } else {
        return color.setRGB(1, 1, 1);
      }
    };


    //******** FUNCTIONS TO CREATE LAYERS FROM CONFIG JSON FILE
    itowns.Fetcher.json('../../../examples/MAM/data/config/layers_settings.json')
      .then(
        (configList) => {
          for (var i in configList) {
            let geometry = configList[i].geometry;
            let protocol = configList[i].source.protocol;
            let config;
            let size;
            let linewidth;

            let addGeometryLayerFromConfig = (config) => {
              if (geometry === "polygons"){
                config.altitude = altitudePoly
              }
              else if (geometry === "lines"){
                config.altitude = altitude;
                if (config.linewidth){
                  linewidth = config.linewidth
                }
                else {
                  linewidth = 50 //default
                };
                config.onMeshCreated = function(result) {
                  result.traverse(function _setLineWidth(mesh) {
                    if (mesh.material) {
                      mesh.material.linewidth = linewidth;
                    }
                  });
                }
              }
              else if (geometry === "points"){
                config.altitude = altitude;
                if (config.size){
                  size = config.size
                }
                else {
                  size = 50 //default
                };
                config.onMeshCreated = function(result) {
                  result.traverse(function _setPointSize(mesh) {
                    if (mesh.material) {
                      mesh.material.size = size;
                    }
                  });
                }
              };
              config.update = itowns.FeatureProcessing.update;
              config.convert = itowns.Feature2Mesh.convert({
                color: new itowns.THREE.Color().setRGB(
                  config.color.red,
                  config.color.green,
                  config.color.blue
                ),
                altitude: config.altitude,
              });
              config.overrideAltitudeInToZero = true;
              config.projection = config.source.projection;
              config.source = config.source;
              let layer = new itowns.GeometryLayer(
                config.id,
                new itowns.THREE.Group(),
                config
              );
              return this.view.addLayer(layer);
            };

            //// If no protocol, set configuration
            if (protocol === undefined) {
              let configuration = configList[i];
              var promise = new Promise(
                function(resolve, reject) {
                  configuration.source = new itowns.FileSource(configList[i].source);
                  resolve(configuration);
                }
              );
              promise.then(addGeometryLayerFromConfig);
            }

            //// If protocol === "wfs", set configuration
            else if (protocol === "wfs") {
              let configuration = configList[i];
              var promise = new Promise(
                function(resolve, reject) {
                  configuration.source = new itowns.WFSSource(configList[i].source);
                  resolve(configuration);
                }
              );
              promise.then(addGeometryLayerFromConfig);
            }
          }
        }
      );
```

### Summary

| Issue | State | iTowns issues |
|:------|:-----:| :-------------|
| Allow to manage layers (*visible, hidden, opacity*) through a graphical interface | Working | https://github.com/iTowns/itowns/issues/1178 |

### Modified files and linked files

| Files | Type | Directory (*relative to UDV-Core directory*) | Status |
|:------|:-----|:--------------------------------------------:|-------:|
| LayerControl.js | JS | */src/Extensions/LayerControls/* | ***Required*** |
| debug.js | JS | */src/Extensions/LayerControls/* | ***Required*** |
| GuiTools.js | JS | */src/Extensions/LayerControls/* | ***Required*** |
| dat.gui.min.js | JS | */src/Extensions/LayerControls/dat.gui/* | ***Required*** |
| Demo.js | JS | */examples/MAM/* | ***Example*** |
| Demo.html | HTML | */examples/MAM/* | ***Example*** |
| example.css | CSS | */examples/MAM/* | ***Required*** |
| Demo.css | CSS | */src/Utils/BaseDemo/css/* | ***Required*** |
| moment.js | JS | */node_modules/moment/* | ***Required*** |
| udvcore.js | JS | */dist/* | ***Required*** |
| three.js | JS | */node_modules/three/build/* | ***Required*** |
| itowns.js| JS | */node_modules/itowns/dist/* | ***Required*** |
| BaseDemo.js | JS | */src/Utils/BaseDemo/js/* | ***Required*** |
| ModuleView.js | JS | */Utils/ModuleView/* | ***Required*** |

### Explanations

As explained in the [iTowns issue](https://github.com/iTowns/itowns/issues/1178), there are iTowns examples about layers Menu creation but no documentation (*however we base our ```LayerControl``` extension on iTowns examples*).

The creation is not really easy but we managed to add a ```LayerControl``` extension, based on the other extensions templates:

```JavaScript
import { ModuleView } from '../../Utils/ModuleView/ModuleView'; //relative path if we're in a src/Category/ModuleName folder
import { GuiTools } from './GuiTools.js'

export class LayerControls extends ModuleView {
    constructor(view, itowns) {
        super() // calling super() is mandatory for subclasses
        this.view = view;
        this.itowns = itowns
    }

    // How to display the view ?
    enableView() {
        this.menuGlobe = new GuiTools('menuDiv', this.view, document.getElementById('viewerDiv'), this.itowns); //calls a method that creates the DOM elements
         for (var layer of this.view.getLayers()) {

                 layer.whenReady.then(  (layer) => {
                   if (layer.visible != undefined)
                   {
                     var gui = debug.GeometryDebug.createGeometryDebugUI(this.menuGlobe.gui, this.view, layer);
                  }
                 });
             }
          }

    // How to close the view ?
    disableView() {
        if (this.menuGlobe !== undefined)
          {
            this.menuGlobe.gui.destroy()
          }; //This methods destroys the DOM elements
    }
}
```

To add this extension, just add this in the main JS file:
```JavaScript
/////// LAYERS CONTROLS
  const layercontrols = new udvcore.LayerControls(baseDemo.view, baseDemo.itowns);
  baseDemo.addModuleView('LayerControls', layercontrols);
```

This extension is based on ```GuiTools.js``` and ```ModuleView```.

* User can click on the Layer Controls button (*see image below*)
Layer_controls

* Then layers can be controlled through the menu that appears:
  * visibility (*Visible checkbox*)
  * opacity (*slider*)
Layer_controls_details

### CODE REVIEW

File: UDV-Core/src/Utils/BaseDemo/js/BaseDemo.js [hunk 1]
```diff
@@ -15,15 +15,10 @@ export class BaseDemo {
...
-        this.extent;  // itowns extent (city limits)
+        this.view; // itowns view (3d scene)
+        this.extent; // itowns extent (city limits)
         this.controls;
-        /**
-         * Object used to manage the 3DTiles layer.
-         *
-         * @type {TilesManager}
-         */
-        this.tilesManager;
```
Comment (jailln) [inline UDV-Core/src/Utils/BaseDemo/js/BaseDemo.js]: Why did you remove ̀tilesManager` ?

File: UDV-Core/src/Utils/BaseDemo/js/BaseDemo.js [hunk 2]
```diff
@@ -514,7 +667,7 @@ export class BaseDemo {
                 throw 'Could not load config file : ' + filePath;
             }
         });
-    }
+    };
```
Comment (jailln) [inline UDV-Core/src/Utils/BaseDemo/js/BaseDemo.js]: unnecessary `;`

File: UDV-Core/src/Utils/BaseDemo/js/BaseDemo.js [hunk 3]
```diff
@@ -479,18 +485,165 @@ export class BaseDemo {
...
+                                        });
+                                }
+                            } else if (geometry === "points") {
+                                config.altitude = altitude;
+                                if (config.size) {
+                                    size = config.size
+                                } else {
+                                    size = 50 //default
+                                }
+                                ;
```
Comment (jailln) [inline UDV-Core/src/Utils/BaseDemo/js/BaseDemo.js]: unnecessary `;`

File: UDV-Core/src/Utils/BaseDemo/js/BaseDemo.js [hunk 4]
```diff
@@ -479,18 +485,165 @@ export class BaseDemo {
...
+                            if (geometry === "polygons") {
+                                config.altitude = altitudePoly
+                            } else if (geometry === "lines") {
+                                config.altitude = altitude;
+                                if (config.linewidth) {
+                                    linewidth = config.linewidth
+                                } else {
+                                    linewidth = 50 //default
+                                }
+                                ;
```
Comment (jailln) [inline UDV-Core/src/Utils/BaseDemo/js/BaseDemo.js]: unnecessary `;`

File: UDV-Core/src/Utils/BaseDemo/js/BaseDemo.js [hunk 5]
```diff
@@ -479,18 +485,165 @@ export class BaseDemo {
...
+
+        //******** FUNCTIONS TO CREATE LAYERS FROM CONFIG JSON FILE
+        itowns.Fetcher.json(
+            '../../../examples/MAM/data/config/layers_settings.json')
+            .then(
+                (configList) => {
+                    for (var i in configList) {
+                        let geometry = configList[i].geometry;
+                        let protocol = configList[i].source.protocol;
+                        let config;
```
Comment (jailln) [inline UDV-Core/src/Utils/BaseDemo/js/BaseDemo.js]: unused local variable

File: UDV-Core/src/Utils/BaseDemo/js/BaseDemo.js [hunk 6]
```diff
@@ -479,18 +485,165 @@ export class BaseDemo {
...
+
+                if (result_altitude) {
+                    z = result_altitude.z;
+                }
+                return z + 10;
+            }
+        };
+
+        // ********* Set the absolute value for polygons layer altitude
+        let altitudePoly = 200;
```
Comment (jailln) [inline UDV-Core/src/Utils/BaseDemo/js/BaseDemo.js]: Why not adding an elevation in a similar fashion than `altitude` ?

File: UDV-Core/src/Utils/BaseDemo/js/BaseDemo.js [hunk 7]
```diff
@@ -479,18 +485,165 @@ export class BaseDemo {
...
+                            let configuration = configList[i];
+                            var promise = new Promise(
+                                function (resolve, reject) {
+                                    configuration.source = new itowns.WFSSource(
+                                        configList[i].source);
+                                    resolve(configuration);
+                                }
+                            );
+                            promise.then(addGeometryLayerFromConfig);
+                        }
```
Comment (jailln) [inline UDV-Core/src/Utils/BaseDemo/js/BaseDemo.js]: this if/else can be simplified by extracting shared code

File: UDV-Core/src/Utils/BaseDemo/js/BaseDemo.js [hunk 8]
```diff
@@ -479,18 +485,165 @@ export class BaseDemo {
...
+                    }
+                }
+            );
```
Comment (jailln) [inline UDV-Core/src/Utils/BaseDemo/js/BaseDemo.js]: The code you added in this method should be in a separate method of BaseDemo (e.g. `addVectorLayers()`) that takes a configuration file as argument. This way, only the users that want these layers in their demo can add them with baseDemo.addVectorLayers(configFile).

File: UDV-Core/src/Utils/BaseDemo/js/BaseDemo.js [hunk 9]
```diff
@@ -15,15 +15,10 @@ export class BaseDemo {
...
+        this.view; // itowns view (3d scene)
+        this.extent; // itowns extent (city limits)
         this.controls;
-        /**
-         * Object used to manage the 3DTiles layer.
-         *
-         * @type {TilesManager}
-         */
-        this.tilesManager;
+        this.itowns = itowns;
```
Comment (jailln) [inline UDV-Core/src/Utils/BaseDemo/js/BaseDemo.js]: You don't need to add this.itowns as an attribute of baseDemo since you set it to a value that is already accessible here.

File: UDV-Core/examples/MAM/Demo.html [hunk 10]
```diff
@@ -1,19 +1,22 @@
...
+        <link rel="stylesheet" href="./example.css">
+
+    </head>
+    <body>
+        <script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jquery/3.1.0/jquery.min.js"></script>
+        <script src="../../node_modules/moment/moment.js"></script>
+        <script src="../../dist/udvcore.js"></script>
+        <script src="../../node_modules/three/build/three.js"></script>
+        <script src="../../node_modules/itowns/dist/itowns.js"></script>
+        <script src="../../node_modules/itowns/dist/debug.js"></script>
```
[the remark this answers is not in the document]
Comment (thomleysens) [reply UDV-Core/examples/MAM/Demo.html]: The Layers Control extension needs ```debug.js``` to run, because it's based on the iTowns GUI menu that is a little bit complicated to manage (*see [this issue] for more details*).

File: UDV-Core/src/Extensions/LayerControls/GuiTools.js [hunk 11]
```diff
@@ -0,0 +1,168 @@
+/**
```
[the remark this answers is not in the document]
Comment (thomleysens) [reply UDV-Core/src/Extensions/LayerControls/GuiTools.js]: Yes ```GuiTools.js``` comes from iTowns. The creation of a layers menu in iTowns is not yet well implemented (*see the [same issue](https://github.com/iTowns/itowns/issues/1178)*) and will probably change. That's why it is copied/pasted and not imported.

File: UDV-Core/src/Extensions/LayerControls/LayerControls.js [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+import { GuiTools } from './GuiTools.js'
+
+export class LayerControls extends ModuleView {
+    constructor(view, itowns) {
+        super() // calling super() is mandatory for subclasses
+        this.view = view;
+        this.itowns = itowns
+    }
+
+    // How to display the view ?
```
[the remark this answers is not in the document]
Comment (thomleysens) [reply UDV-Core/src/Extensions/LayerControls/LayerControls.js]: What did you not understand ? 
If it's related to ```this.itowns = itowns```, this line helps to build the menu and manage required global variable.

File: UDV-Core/src/Extensions/LayerControls/LayerControls.js [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+    enableView() {
+        this.menuGlobe = new GuiTools('menuDiv', this.view, document.getElementById('viewerDiv'), this.itowns); //calls a method that creates the DOM elements
```
[the remark this answers is not in the document]
Comment (thomleysens) [reply UDV-Core/src/Extensions/LayerControls/LayerControls.js]: You are absolutely right. Will be changed in the next commit.

**Comment by EricBoix** [comment]
GUI menus where never updated/integrated to UD-Viz "standard".
Closing.
